### PR TITLE
Adds Network Resource Uploading for admins.

### PR DESCRIPTION
**Pull request by Zumorica**
## About the PR 
~~Requires submodule update to https://github.com/space-wizards/RobustToolbox/commit/aa5bcefaf25d824f75ca6373c9acefca26d136e3~~
~~Requires https://github.com/space-wizards/RobustToolbox/pull/2590 and a submodule update~~.

Admins with +QUERY can now upload any sort of resource to the server and clients.

By default, all uploaded resources are logged to the server's database, and automatically deleted every 30 days (auto-deletion happens on server init)
There are CVars to control both of these behaviors, uploaded file size limit, and whether resource uploading is enabled at all.

This is my first time doing database stuff, so please review that part thoroughly. I've tested it on both SQLite and PostgreSQL and it seems to work fine. 

**Screenshots**

https://user-images.githubusercontent.com/6766154/155848163-a8ace88d-26b4-4e54-bf40-5fd7afb43600.mp4

### CODE REVIEW

File: Content.Server.Database/Model.cs [hunk 1]
```diff
@@ -438,4 +439,19 @@ public sealed class ServerRoleUnban
...
+        [Key, DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public int Id { get; set; }
+
+        [Required] public DateTime Date { get; set; }
+
+        [Required] public Guid UserId { get; set; }
+
+        [Required] public string Path { get; set; } = string.Empty;
+
+        [Required] public byte[] Data { get; set; } = default!;
```
Comment (PJB3005) [inline Content.Server.Database/Model.cs:455]: `[Required]` isn't necessary here.

File: Content.Server/Administration/NetworkResourceManager.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        // Do not allow uploading any new resources if it has been disabled.
+        // Note: Any resources uploaded before being disabled will still be kept and sent.
+        if (!Enabled)
+            return;
+
+        if (!_playerManager.TryGetSessionByChannel(msg.MsgChannel, out var session))
+            return;
+
+        // +QUERY only for now.
+        if (!_adminManager.IsAdmin(session) || !_adminManager.HasAdminFlag(session, AdminFlags.Query))
```
Comment (PJB3005) [inline Content.Server/Administration/NetworkResourceManager.cs:51]: `HasAdminFlag` already checks whether they're admin.

```suggestion
        if (!_adminManager.HasAdminFlag(session, AdminFlags.Query))
```

File: Content.Server/Database/ServerDbBase.cs [hunk 3]
```diff
@@ -754,6 +754,33 @@ public async Task RemoveFromWhitelistAsync(NetUserId player)
...
+        {
+            await using var db = await GetDb();
+
+            var today = DateTime.Now;
+
+            await foreach (var log in db.DbContext.UploadedResourceLog.AsAsyncEnumerable())
+            {
+                if ((today - log.Date).TotalDays >= days)
+                    db.DbContext.Remove(log);
+            }
```
Comment (PJB3005) [inline Content.Server/Database/ServerDbBase.cs:777]: This is pulling in the whole table. You should properly filter the query with a `.Where` to only get things that are too old.

File: Content.Shared/Administration/SharedNetworkResourceManager.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        ResourceManager.AddRoot(Prefix, this);
+    }
+
+    protected abstract void ResourceUploadMsg(NetworkResourceUploadMessage msg);
+
+    public bool TryGetFile(ResourcePath relPath, [NotNullWhen(true)] out Stream? stream)
+    {
+        byte[]? data;
+
+        lock(Files)
```
Comment (PJB3005) [inline Content.Shared/Administration/SharedNetworkResourceManager.cs:42]: The locking strategy here should be a `ReaderWriterLockSlim`, not a `lock` mutex.

File: Content.Shared/Administration/SharedNetworkResourceManager.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+using Robust.Shared.Network;
+using Robust.Shared.Utility;
+
+namespace Content.Shared.Administration;
+
+/// <summary>
+///     Manager that allows resources to be added at runtime by admins.
+///     They will be sent to all clients automatically.
+/// </summary>
+public abstract class SharedNetworkResourceManager : IContentRoot
```
Comment (PJB3005) [inline Content.Shared/Administration/SharedNetworkResourceManager.cs:13]: I really do not like that this manager directly implements `IContentRoot`. Makes for very messy object lifetimes.

**Review by PJB3005 — COMMENTED** [review]
I am gonna merge #6327 before this since it's been sitting too long and I need to do these migrations in order to preserve my sanity.

Migrations need to be re-created from 6b66c7db8a23c96ce15c6190e5340cd4e7f990c2. I'll do that myself though.

File: Content.Server/Database/ServerDbBase.cs (hunk 3 above)
Comment (Zumorica) [reply Content.Server/Database/ServerDbBase.cs]: Done.

File: Content.Shared/Administration/SharedNetworkResourceManager.cs (hunk 4 above)
Comment (Zumorica) [reply Content.Shared/Administration/SharedNetworkResourceManager.cs]: Done. I hope it's correct now

File: Content.Shared/Administration/SharedNetworkResourceManager.cs (hunk 5 above)
Comment (Zumorica) [reply Content.Shared/Administration/SharedNetworkResourceManager.cs:14]: I think I'll move this `IContentRoot` implementation to engine, given it's generic enough that multiple things or games may want to make use of it.

File: Content.Shared/Administration/SharedNetworkResourceManager.cs (hunk 4 above)
Comment (Zumorica) [reply Content.Shared/Administration/SharedNetworkResourceManager.cs]: Also this requires a new entry in the sandbox whitelist, but I'm gonna move the impl of `IContentRoot` to engine so not a big deal for now

File: Content.Server.Database/Model.cs (hunk 1 above)
Comment (Zumorica) [reply Content.Server.Database/Model.cs]: Done.

File: Content.Shared/Administration/SharedNetworkResourceManager.cs (hunk 5 above)
Comment (Zumorica) [reply Content.Shared/Administration/SharedNetworkResourceManager.cs]: I've done that now in https://github.com/space-wizards/RobustToolbox/pull/2590, is it better now?

**Review by mirrorcult — APPROVED** [review]

**Review by PJB3005 — COMMENTED** [review]
I'm gonna re-order the migration on top of master for you.

**Pull request merged**